File: sim/report.go
package sim

import (
	"time"

	"economy/citizen"
	"economy/country"
)

// PriceQuote represents top-of-book market quote for one symbol at one tick.
type PriceQuote struct {
	Symbol string  `json:"symbol"`
	Mid    float64 `json:"mid"`
	Bid    float64 `json:"bid"`
	Ask    float64 `json:"ask"`
	Spread float64 `json:"spread"`
	HasMid bool    `json:"has_mid"`
}

// Trade represents an executed match between taker and maker on a specific symbol.
type Trade struct {
	Tick         int     `json:"tick"`
	Symbol       string  `json:"symbol"`
	TakerAgentID string  `json:"taker_id"`
	MakerAgentID string  `json:"maker_id"`
	Side         string  `json:"side"` // "BUY" or "SELL"
	Price        float64 `json:"price"`
	Quantity     float64 `json:"quantity"`
}

// AccountSnapshot captures an account's financial telemetry at a tick.
type AccountSnapshot struct {
	AgentID     string  `json:"agent_id"`
	Cash        float64 `json:"cash"`
	Equity      float64 `json:"equity"`
	MarginRatio float64 `json:"margin_ratio"`
}

// TickReport is returned by Simulation.Step(), summarizing all activity during that tick.
type TickReport struct {
	Tick      int                    `json:"tick"`
	Timestamp time.Time              `json:"timestamp"`
	Prices    map[string]PriceQuote  `json:"prices"`
	Trades    []Trade                `json:"trades"`
	Events    []Event                `json:"events"`
	Accounts  []AccountSnapshot      `json:"accounts"`
	Citizen   citizen.CitizenReport  `json:"citizen"`
	National  country.NationalReport `json:"national"`
}

File: src/sim/report.rs
use std::collections::HashMap;
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

use crate::citizen::CitizenReport;
use crate::country::NationalReport;
use super::events::Event;

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct PriceQuote {
    pub symbol: String,
    pub mid: f64,
    pub bid: f64,
    pub ask: f64,
    pub spread: f64,
    pub has_mid: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Trade {
    pub tick: usize,
    pub symbol: String,
    pub taker_id: String,
    pub maker_id: String,
    pub side: String,
    pub price: f64,
    pub quantity: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AccountSnapshot {
    pub agent_id: String,
    pub cash: f64,
    pub equity: f64,
    pub margin_ratio: Option<f64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TickReport {
    pub tick: usize,
    pub timestamp: DateTime<Utc>,
    pub prices: HashMap<String, PriceQuote>,
    pub trades: Vec<Trade>,
    pub events: Vec<Event>,
    pub accounts: Vec<AccountSnapshot>,
    pub citizen: CitizenReport,
    pub national: NationalReport,
}
